pkg/linter/linters/fetch.go

```diff
@@ -0,0 +1,401 @@
+	Branch         string
+	Ref            string
+	Tag            string
+	ExpectedCommit string
+}
+
+// Holds compiled regex patterns for a specific package
+type packagePatterns struct {
+	exactVersionURL   *regexp.Regexp
+	anyVersionURL     *regexp.Regexp
+	exactVersionTag   *regexp.Regexp
+	packageVersionTag *regexp.Regexp
+}
+
+// Holds extracted pipeline data
+type sourceData struct {
+	fetchURLs   []string
+	gitTags     []string
+	gitBranches []gitRefInfo
+}
+
+// Handles validation logic with pre-compiled patterns
+type validator struct {
+	pkg      packageInfo
+	patterns *packagePatterns
+}
+
+// Holds package metadata
+type packageInfo struct {
+	name    string
+	version string
+}
+
+var (
+	anyTemplateRegex     = regexp.MustCompile(`\$\{\{[^}]+\}\}`)
+	versionTemplateRegex = regexp.MustCompile(`\$\{\{\s*(package\.(version|full-version)|vars\.[^}]*\bversion\b[^}]*)[^}]*\}\}`)
+	shaRegex             = regexp.MustCompile(`^[0-9a-fA-F]{12,40}$`)
+)
+
+func (s sourceData) isEmpty() bool {
+	return len(s.fetchURLs) == 0 && len(s.gitTags) == 0 && len(s.gitBranches) == 0
+}
+
+// Creates compiled regex patterns for package-specific validation
+func buildPackagePatterns(packageName, packageVersion string) *packagePatterns {
+	if packageName == "" || packageVersion == "" {
+		return nil
+	}
+
+	escName := regexp.QuoteMeta(packageName)
+	escVer := regexp.QuoteMeta(packageVersion)
+
+	return &packagePatterns{
+		exactVersionURL: regexp.MustCompile(
+			`(?:^|/)` + escName +
+				`(?:[-_]v?` + escVer + `(?:\.tar\.(?:gz|bz2|xz)|\.zip|\.tgz|\.tbz2|\.txz)` +
+				`|/v?` + escVer + `/)`),
+		anyVersionURL: regexp.MustCompile(
+			`(?:^|/)` + escName +
+				`(?:[-_]v?\d+\.\d+(?:\.\d+)?(?:[-+][\w\d.]+)?` +
+				`|/v?\d+\.\d+(?:\.\d+)?(?:[-+][\w\d.]+)?/)`),
+		exactVersionTag:   regexp.MustCompile(`^v?` + escVer + `$`),
+		packageVersionTag: regexp.MustCompile(`^` + escName + `[-_]v?\d+\.\d+(?:\.\d+)?(?:[-+][\w\d.]+)?$`),
+	}
+}
+
+// Fetch sources and git data from raw YAML before template substitution
+func extractRawPipelineData(root *yaml.Node) sourceData {
+	var fetchSources []string
+	var gitTags []string
+	var gitBranchRefs []gitRefInfo
+
+	if root == nil {
+		return sourceData{
+			fetchURLs:   fetchSources,
+			gitTags:     gitTags,
+			gitBranches: gitBranchRefs,
+		}
+	}
+
+	// Unwrap DocumentNode to get content
+	if root.Kind == yaml.DocumentNode && len(root.Content) > 0 {
+		root = root.Content[0]
+	}
+
+	findValue := func(mapping *yaml.Node, key string) *yaml.Node {
+		if mapping.Kind != yaml.MappingNode {
+			return nil
+		}
+		for i := 0; i < len(mapping.Content); i += 2 {
+			if i+1 < len(mapping.Content) && mapping.Content[i].Value == key {
+				return mapping.Content[i+1]
+			}
+		}
+		return nil
+	}
+
+	getString := func(mapping *yaml.Node, key string) string {
+		if node := findValue(mapping, key); node != nil && node.Kind == yaml.ScalarNode {
+			return node.Value
+		}
+		return ""
+	}
+
+	var processPipelines func(*yaml.Node)
```

Comment:
Would it make sense to have `findValue`, `getString`, and `processPipelines` exist as separate functions for maintainability and to improve future testing?